#177 User Location Tracking for schedules

refatK opened the pull request:
<!--
NOTE: Please set, 
- reviewers for the PR
- assign yourself in "Assignees"
- "pull request" label for "Labels"
- set the correct "Milestone",
- set the "Linked Issues" to the issue this PR would close, if any

in the menu on the right
-->

### Related Issues
Closes #177

### Important Notes
- ~I created a test toggle in the [Constants.java](https://github.com/Landry333/Big-Owl/pull/238/files#diff-0ca77fa01f082b3a89e2f6f5c76cb4a3aec7632da690a38202fbce09504e7263) to make it easy to turn on/off the testing aspect of this code. This was necessary as this feature relies on #179 and #15 to work fully. I had to create some test only code so that everyone could actually test the feature. By using a feature toggle, it will be easy to put off this test code if need be, and it will also make it must easier to identify what should be removed when the related features are implemented~

- ~As a result of this test code, there are some sonar cloud errors, but these are guaranteed to be removed when the test code is removed, therefore the warnings that are left can be ignored. I have made sure to fix sonarcloud code smells related to code that will not be removed, although you can bring up any if it seems I missed one not related to test code~

- testing will be done later as some of our issues depend on it, and we should get the code in as so as possible so that we don't delay other issues

### Description
I have added the location checker will check the user is in the correct location based on a schedule.

When called the users location is initially checked. It is then checked for up to 30 mins allowing the user time to enter the scheduled location if needed. Once the user has been detected as being in the location, the tracking is stopped.

Android geofencing API is used to accomplish this. Everything runs in the background so the user doesn't have to keep the app open for the location check. The geofence api only cares about wether the person gets into and out of the radius of a selected location and does not constantly notify the app if the user is changing location but not changing their state of being within or outside the radius of a specific location.

I have also created a permissions helper to make permission requests much simpler. I have used it to request location permissions which are necessary to ask for to allow this feature to work. I decided to ask for the location permissions once a person accepts a schedule request as I felt that was an appropriate time to warn the user as to why location tracking will be needed. If you guys think there is a better place to put it, I can do that.

### How to check PR

#### Checking permissions request:

- Login as testuser8
- Go to any schedule request (SUPERVISED GROUP -> click testuser1 group -> SCHEDULES -> choose a schedule)
- Accept any request (if there are none to accept, just reject and accept again)
- Follow and accept all location permission dialogues

You should get a PERMISSIONS_GRANTED toast if successful.

You should also try cases where you deny, and also check if you can get permissions even after denial.

If you grant full permission check that you aren't warned as well. If a user has already given the permissions, there is no reason to warn them.


#### Checking location tracking:

To test this feature optimally, you will need to use fake locations so that you can easiy change your location in and out of expected areas, there are 2 methods of doing this.

1. On emulator: The emulator has a native option to mock your location [explained here](https://stackoverflow.com/a/45098850)
2. On real device: Install [this app](https://play.google.com/store/apps/details?id=com.lexa.fakegps&hl=en_CA&gl=US) and in Android Developer options of your android device set `Select mock location app` to the app you just installed "Fake GPS".

You can also change the hyper-parameters `DEFAULT_TRACKING_EXPIRE_TIME_MILLIS` and `DEFAULT_MAX_NOTIFY_DELAY_MILLIS` and also `DEFAULT_TRACKING_RADIUS_METERS` in the class `ScheduledLocationTrackingManager` to make testing easier (instead of waiting 30 minutes for the "never got to correct location case" case, you can set it to a lower time for example). You might want to set `DEFAULT_MAX_NOTIFY_DELAY_MILLIS` to speed up the time it takes for the app to detect location changes.

- Make sure to re-install the app since new `BroadcastReceiver`s were added to the app which only get set on installation
- Make sure the app has full location permissions and android location services is enabled
- Login as testuser8
- Go to `LOCATION CHECK TEST` schedule request (SUPERVISED GROUP -> click testuser1 group -> SCHEDULES -> choose `LOCATION CHECK TEST`), and make sure it is accepted.
- Go back to homepage
- Open [this page](https://console.firebase.google.com/u/0/project/bigowlapp-ca8da/firestore/data~2Fschedules~2Ff5XyMr8LdFgmO5TyqTG5) on a browser, and set a `startTime` of the database document to a time that is soon to come (must be at least one minute after your current time)
 - - You can also change the `location` to test with whatever location you please
- Load/Reload the homepage of the app before this time and wait for the time to come
- Check the firebase document you opened previously and the field for user `8YQB2MXQ4bUGLyvBejUEhnYHtK83` should change regarding the `scheduleLocated` and `authenticationTime` fields

Cases that should be tested:

1. Already in location: If you are already in a location: the  app should detect that your are in the location and store `CORRECT_LOCATION` TO THE `scheduleLocated` field of the user. All location detection should end and `LocationTrackingExpiredAlarmReceiver` should never be called at the expire time defined by `DEFAULT_TRACKING_EXPIRE_TIME_MILLIS`

2. Never gets to location: If you are not in the correct location, the app should initially set `scheduleLocated` to `WRONG_LOCATION`, but location detection should still continue and only end until the expire time defined by `DEFAULT_TRACKING_EXPIRE_TIME_MILLIS`

3. Gets to location eventually: Same as above, but if you then enter the location before the expire time is reached, the database should eventually change the `scheduleLocated` to `CORRECT_LOCATION` (might have a couple minutes delay), at which point it should behave as described in case 1.

<del>

- Make sure the app has full location permissions  and android location services is enabled
- Login as testuser2
- go to set schedule
- You must set a *Title* and the *location* and then click the big "Test Only" button

Based on your devices location relative to the inputted location, a toast should pop up indicating whether you were indeed within the radius of the location or not. The database should also update the schedule with id `4laPgh1xNIy8CDpnohDV` for testUser2 (who's id is `vjsRjs4WtieuBd7qxeFPi8jo2qr1`) for scheduleLocated field. This schedule is simply being used to test, the code is already set to update the actual appropriate schedule. Logs will also be useful for more details about what is going on.

You might want to test with the emulator as you can spoof locations with it as well to more robustly test the feature. Such as the case where the user is detected as outside the location first, but then gets marked as being there when in the location.

</del>

### Important Files
- [LocationBroadcastReceiver.java](https://github.com/Landry333/Big-Owl/pull/238/files#diff-4546903a7a8d28b7b235f1a2659b7103b7ee108c3f9fce2e6d8dd617a8fc6da2) - Handles what happens when the user exits/enters a location within a range
- [ScheduledLocationTrackingManager.java](https://github.com/Landry333/Big-Owl/pull/238/files#diff-f5b81adc3c75f53025e6f431075f66084a5f9ab3277c4fee69cb4a8a82a47453) - Generates location tracking geofences. You can modify some hyperparameters here if needed
- [PermissionsHelper.java](https://github.com/Landry333/Big-Owl/pull/238/files#diff-3cbd8c32e2939b93fb71846a53f9219570ac9dfdcd51cdb7914d36bf76d9442c) - manages the handling of permission requests
- [PeriodicLocationCheckAlarmManager.java](https://github.com/Landry333/Big-Owl/pull/238/files#diff-fdf77b355b98fe47ccd8930e75516d933952e183a7108176c5863f032357209e) - Adds recurring alarm to force location calculation when user location is being tracked.
- [LocationTrackingExpiredAlarmManager.java](https://github.com/Landry333/Big-Owl/pull/238/files#diff-384d440bb0fff5fb14fc62436fccbba4117d1d6520bc78762dfeea66e1fbd564) - created and Alarm for when location tracking has expired as a result of user not getting to the location on time.
- [LocationTrackingExpiredAlarmReceiver.java](https://github.com/Landry333/Big-Owl/pull/238/files#diff-3eab5313967d0cb3944835106ce04b3b8fd8c79d3b8503b51a1e431848f51487) - What happens when location tracking has expired as a result of user not getting to the location on time.

ShawnGuerra-Bautista committed:
[#177] Added permissions for background location access

Co-authored-by: RefatK <refat-17@hotmail.com>

refatK committed:
[#177] Added incomplete LocationBroadcastReceiver which will handle when user leaves/enters a scheduled location

refatK committed:
Merge branches '#177-location' and 'staging' of https://github.com/Landry333/Big-Owl into #177-location

refatK committed:
[#177] Added ScheduledLocationTrackingManager which will be used to create location tracking requests

refatK committed:
[#177] Add addNewLocationToTrack function to track locations

refatK committed:
[#177] Added code that will be used to check the ScheduledLocationTrackingManager works

refatK committed:
[#177] Added Fine Location Permissions and added some minor fixes. Seems to work at this point

refatK committed:
[#177] Stop tracking when user successfully enters the vicinity of a desired location

refatK committed:
[#177] Location tracking now uses Schedule objects and uses Schedule Id for identification

refatK committed:
[#177] Fix Location result taking very long initially, and only watch for initial exit

refatK committed:
Merge branch 'staging' of https://github.com/Landry333/Big-Owl into #177-location

 Conflicts:
	app/build.gradle

refatK committed:
[#177] Implement case where user got to location and their location is set as correct

refatK committed:
[#177] Refactor: Extract Function - extracted LocationBroadcastReceiver.setUserLocatedStatus for location update for better reuse and readability

refatK committed:
[#177] Refactor: Rename Method - make it more clear this involves a database update

refatK committed:
[#177] User will now be correctly idetified based on situation

refatK committed:
[#177] made sure to prevent possible duplicate database calls

refatK committed:
[#177] Fix Crash related to permissions and small code cleanup

refatK committed:
[#177] Refactor: Extract Method - Move permission handling to a method to make onRequestPermissionsResult less bloated in case of further added permission calls

refatK committed:
[#177] Refactor: Combine Functions into Class: created PermissionsHelper to make permission handling more reusable, and debloat activity

refatK committed:
[#177] Refactor: Replace Loop with Pipeline: emphsize that we are simply filtering out granted permissions

refatK committed:
[#177] Refactor: Extract Function: make it more clear what the filter is doing

refatK committed:
[#177] make the PermissionsHelper generic, and not just for location

refatK committed:
[#177] Further permission request cleanup

refatK committed:
[#177] Move location permission warning/request to a more logical location (user accepts a schedule) and other fixes

refatK committed:
[#177] Fixing bugs related to gefence removal and some clean up

refatK committed:
Merge branches '#177-location' and 'staging' of https://github.com/Landry333/Big-Owl into #177-location

refatK committed:
[#177] Fix test that broke due to missing permissions

refatK committed:
[#177] Quality/Refactor: Replace magic literal - Use static variable for clarity as well as to remove code smell

refatK committed:
[#177] Quality/Refactor: Rename field, Rename method - Make field use camelCase instead

refatK committed:
[#177] Quality: Fixed some more sonarcloud issues

refatK committed:
[#177] Fix permissions request for location

refatK committed:
[#177] Refactor: Rename Field - used much more clear request code names and also added documentation for less confusion

refatK committed:
[#177] Refactor: Extract Method - created method requestBackgroundLocationPermissions() to remove dupicated code

refatK committed:
[#177] Use better justification with styling to make the reason for background location more clear.

refatK committed:
[#177] Quality/Refactor: Extract Method - reduce duplication by reusing overloaded functions instead

refatK committed:
[#177] Quality: added documentation and changed visibility to make PermissionsHelper easier to use

refatK committed:
[#177] Refactor: Rename Method: Use a better name for function that describe the action of requesting location permissions, not permissions in general

refatK committed:
[#177] Refactor: Remove Dead Code - removed unused private methods that had no actual use

refatK committed:
[#177] Quality: fix bug pointed out by SonarCloud

refatK committed:
[#177] Quality: Remove redundant code. The return type of buildRequestToTrack() is already Task<Void>, so there's no need to specify a onSuccessTask that returns a Task<Void>.
Also wrapped some lgs useful for testing but should be removed later.

refatK committed:
[#177] Quality/Refactor: Extract or Move Method; Move location request code from ScheduleViewRespondActivity to PermissionsHelper to increase cohesion and reduce ScheduleViewRespondActivity dependency to Manifest. Also allows for better reuse.

KhangZhengNg commented on app/src/main/java/com/example/bigowlapp/model/Attendance.java:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    private boolean authenticated = false;
+    private boolean authAttemptedMethod1 = false;  // Authentication with first method
+    private boolean authAttemptedMethod2 = false;  // Authentication with second method
```
Provide better naming and maybe briefly comment what they do

refatK replied:
Since these variables are actively being worked with on a different branch, I'll leave it as is for now.

Although I agree with your point.

KhangZhengNg commented on app/src/main/java/com/example/bigowlapp/service/LocationBroadcastReceiver.java:
```diff
@@ -0,0 +1,125 @@
+                .distinct()
+                .collect(Collectors.toList());
+        Log.e(TAG, "SCHEDULE IDS THAT TRIGGERED THIS ARE: " + geofenceIdList);
+        List<String> testOnlyIdList = Collections.singletonList("4laPgh1xNIy8CDpnohDV");
```
Add a TODO tag since it will be removed in the future

refatK replied:
Nice catch, this should have been wrapped in the Test toggle as well

refatK replied:
don now

KhangZhengNg commented on app/src/main/java/com/example/bigowlapp/utils/PermissionsHelper.java:
```diff
@@ -0,0 +1,176 @@
+     * all possible scenarios regarding location/background location request
+     */
+    public void requestLocationAndBackgroundLocationPermissions() {
+        String reason = "Location detection is required to check whether you have arrived to a scheduled location." +
+                "\n\nIf not provided, the system will assume you have not arrived to the location.";
```
Store this String in Strings.xml

refatK replied:
done now

KhangZhengNg reviewed (changes requested):
Very good efforts on implementing Geofencing API

refatK committed:
[#177] Correctly mark test code as indicated by review

refatK committed:
[#177] move location justification string to strings.xml as indicated by review

refatK committed:
[#177] Add timestamp for time of detection

refatK committed:
Merge branches '#177-location' and 'staging' of https://github.com/Landry333/Big-Owl into #177-location

 Conflicts:
	app/src/main/res/values/strings.xml

refatK committed:
Merge branches '#177-location' and 'staging' of https://github.com/Landry333/Big-Owl into #177-location

 Conflicts:
	app/src/main/AndroidManifest.xml

refatK committed:
[#177] remove test toggle and code related to it

refatK committed:
[#177] implemented location tracking with actual scheduled alarms

ShawnGuerra-Bautista commented on app/build.gradle:
```diff
@@ -44,6 +44,12 @@ android {
+    lintOptions {
+        disable "UnsafeExperimentalUsageError",
+                "UnsafeExperimentalUsageWarning",
+                "LogNotTimber"
```
Nice change; we don't need to worry about the timber warnings that are useless to us.

ShawnGuerra-Bautista commented on app/src/main/java/com/example/bigowlapp/model/Attendance.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    private boolean authenticated = false;
+    private boolean authAttemptedMethod1 = false;  // Authentication with first method
+    private boolean authAttemptedMethod2 = false;  // Authentication with second method
```
This comment is not requesting a change; I'm only going to give my feedback concerning these lines.

I know this is an issue concerning @Landry333, but I think we should only use a single method of authentication. Also, the boolean called "authenticated" should be "isAuthenticated" as that is a much more precise name and it indirectly states that this field is a boolean.

refatK replied:
I agree with these points and will discuss with @Landry333

ShawnGuerra-Bautista commented on app/src/main/java/com/example/bigowlapp/service/MemberScheduleAlarmReceiver.java:
```diff
@@ -22,9 +24,13 @@
+        ScheduledLocationTrackingManager locationTrackingManager = new ScheduledLocationTrackingManager(context);
+        // TODO: figure what to do with success/failure, or remove them
+        locationTrackingManager.addNewScheduledLocationToTrack(activatedSchedule)
+                .addOnSuccessListener(aVoid -> Log.e("BigOwl", "LOCATION TRACKING STARTED"))
+                .addOnFailureListener(e -> Log.e("BigOwl", e.getMessage()));
```
I saw the comment, and I believe it's either remove them, or present a toast instead of logging.

Though, this should only be done before merging into staging; it could be useful for testing the PR code.

refatK replied:
- [x] Completed?

ShawnGuerra-Bautista commented on app/src/main/java/com/example/bigowlapp/utils/Constants.java:
```diff
@@ -1,6 +1,11 @@
 
 public interface Constants {
+    // constants for time conversion
+    int MINUTE_TO_SECONDS = 60;
+    int SECONDS_TO_MILLISECONDS = 1000;
+    int MINUTE_TO_MILLISECONDS = MINUTE_TO_SECONDS * SECONDS_TO_MILLISECONDS;
```
Those are nice.

ShawnGuerra-Bautista commented on app/src/main/java/com/example/bigowlapp/utils/PermissionsHelper.java:
```diff
@@ -0,0 +1,175 @@
+ * This helper can be used when needing to handle calls for permissions
+ * without bloating an Android UI class, and to reduced repeated code.
+ */
+public class PermissionsHelper {
```
Really cool class to handle permissions.

Though, I have a question as I don't really understand some of the code. Are most of these methods flexible? As in, if I'm requesting for another permission other than location, will this class help?

refatK replied:
Yep, I created some methods for location specifically as background location has it's own complexity, but everything else can use `requestMissingPermissions()`.

Anything public in `PermissionsHelper` can be relevant for anyone to use.

If there's a public method that shouldn't be public, just tell me.

ShawnGuerra-Bautista commented on app/src/main/java/com/example/bigowlapp/utils/ScheduledLocationTrackingManager.java:
```diff
@@ -0,0 +1,81 @@
+    private static final int DEFAULT_TRACKING_RADIUS_METERS = 300;
+    private static final long DEFAULT_TRACKING_TIME_MILLISECONDS = 30L * Constants.MINUTE_TO_MILLISECONDS;
+    // use 0 for instant response, and allow delay for better battery life
+    private static final int DEFAULT_MAX_NOTIFY_DELAY_MILLISECONDS = 5 * Constants.MINUTE_TO_MILLISECONDS;
```
I tested your code, and, for some reason, the geofence feature does not notify/check every 5 minutes if a user is not at the desired location.

Although, I'll give you an interesting find: it only re-notified when I opened the Google Maps app, so I guess you have to "wake" up the location feature.

refatK replied:
I've noticed that too. Will try to use a recurring alarm to fix this issue.

- [x] Completed?

refatK replied:
Seems like the recurring alarm idea worked.

ShawnGuerra-Bautista commented on app/src/main/java/com/example/bigowlapp/utils/ScheduledLocationTrackingManager.java:
```diff
@@ -0,0 +1,81 @@
+        }
+
+        Intent intent = new Intent(context, LocationBroadcastReceiver.class);
+        geofencePendingIntent = PendingIntent.getBroadcast(context, 0, intent,
```
Refactor: Replace magic literal. 

The number 0 should have a meaning in this context, here it's unclear.

refatK replied:
- [x] Completed?

ShawnGuerra-Bautista commented on app/src/main/java/com/example/bigowlapp/service/LocationBroadcastReceiver.java:
```diff
@@ -0,0 +1,99 @@
+public class LocationBroadcastReceiver extends BroadcastReceiver {
+    private static final String TAG = LocationBroadcastReceiver.class.getName();
+
+    RepositoryFacade repositoryFacade;
```
I believe this field should be private.

refatK replied:
- [x] Completed?

ShawnGuerra-Bautista commented on app/src/main/java/com/example/bigowlapp/service/LocationBroadcastReceiver.java:
```diff
@@ -0,0 +1,99 @@
+
+            // User was successfully detected in desired location, so no more tracking needed
+            this.removeLocationTracking(context, geofenceIdList)
+                    .addOnSuccessListener(aVoid ->
+                            Log.e(TAG, "Entered LOCATION TRACKING SUCCESSFULLY REMOVED"))
+                    .addOnFailureListener(e ->
+                            Log.e(TAG, "FAILED TO REMOVE LOCATIONS"));
```
Make sure to remove them before merging, but keep it for now since it will be useful when testing for PR review.

refatK replied:
- [x] Completed?

ShawnGuerra-Bautista commented on app/src/main/java/com/example/bigowlapp/service/LocationBroadcastReceiver.java:
```diff
@@ -0,0 +1,99 @@
+            return;
+        }
+
+        // valuable data from geofence result
+        // TODO: if not needed remove, else uncomment for users current location
+        //  Location currentUserLocation = geofencingEvent.getTriggeringLocation();
```
Is the valuable data only from the code that you commented out, and is it that useful?

If it isn't used, then it should be removed.

refatK replied:
Not for me, but might be for the authentication task, so I'll leave  it for authentication task to be used or  removed

refatK committed:
[#177] Refactor: Replace Magic Literal; made it clear the 0 was referring to a request code

refatK committed:
[#177] Quality: make a class attribute private as requested in review and final

refatK committed:
[#177] create an AlarmManger that will do a recurring location check.

refatK committed:
[#177] add the handling of the recurring alarm by adding the location check code

refatK committed:
[#177] get the recurring location check to work

refatK committed:
[#177] cancel location checking once a person has been located

refatK committed:
[#177] Refactor: Extract Method, Replace Inline Code with Function Call; added a function to reduce duplication of inline code

refatK committed:
[#177] Fix geofence not being called and minor cleanup

refatK committed:
[#177] Add CancelPeriodicLocationCheckAlarmReceiver to make sure the location eventually get cancelled.

ShawnGuerra-Bautista commented on app/src/main/java/com/example/bigowlapp/utils/PeriodicLocationCheckAlarmManager.java:
```diff
@@ -0,0 +1,77 @@
+        return PendingIntent.getBroadcast(context, REQUEST_CODE_CANCEL, intent, NO_FLAG);
+    }
+
+    public static class PeriodicLocationCheckAlarmReceiver extends BroadcastReceiver {
+        @Override
+        public void onReceive(Context context, Intent intent) {
+            FusedLocationProviderClient locationClient = LocationServices.getFusedLocationProviderClient(context);
+            if (ContextCompat.checkSelfPermission(context, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
+                return;
+            }
+            locationClient.getCurrentLocation(LocationRequest.PRIORITY_HIGH_ACCURACY, null);
+        }
+    }
+
+    public static class CancelPeriodicLocationCheckAlarmReceiver extends BroadcastReceiver {
+        @Override
+        public void onReceive(Context context, Intent intent) {
+            PeriodicLocationCheckAlarmManager locationCheckAlarmManager = new PeriodicLocationCheckAlarmManager(context);
+            locationCheckAlarmManager.cancelPeriodicLocationCheck();
+        }
+    }
+}
```
Use an if statement instead of creating 2 different classes; it could be merged into 1 class.

KennanHoa replied:
I'll be using the CancelPeriodicLocationCheckAlarmReceiver to make the missed schedule notification so it should be moved to a separate class on its own instead.

refatK replied:
I guess I'll make the inner class an outer class and rename it to reflect its more broad use then, although @ShawnGuerra-Bautista initial point was right if it was going to simply cancel the repeating alarm, which was what I made it for at first.

refatK committed:
[#177] fix crash when denying permissions

KennanHoa reviewed (approved):
Feature work as intended.
- Database changed depending on the location of the user at the scheduled time.
- Application stopped searching the location if the user is at the right location or is still at the wrong location after 30 minutes.

Clean and well organized codes.

refatK committed:
[#177] Refactor: Extract class, Move Method: Created a new class for managing the case location tracking expires so that we can do things when this happen.

Also extracted class and moved methods to LocationTrackingExpiredAlarmManager for this new receiver.

refatK committed:
[#177] Minor fixes and changes after refactor

refatK committed:
[#177] Quality: address final changes and do minor clean up to prepare for merge.

refatK committed:
[#177] Quality: remove unused import

refatK committed:
[#177] If user momentarily goes out of geofence, there is no need to mark them as being in the incorrect location.

refatK committed:
[#177] Refactor: Rename attribute; Use millis instead of milliseconds in variable names to make them more concise, and follow the standard naming used by Java and Android

KhangZhengNg reviewed (approved):
Good fixes. More tests needed, but we need these key parts for others to work on.

refatK commented:
> Good fixes. More tests needed, but we need these key parts for others to work on.

Totally agree!

refatK committed:
[#177] Remove another assert that leads to flaky tests
